### DISCUSSION TITLE
how to (properly) get the address of the server to which the current client connects?

### DISCUSSION THREAD

**Discussion by marcin-koziol** [zauberzeug/nicegui#2213] [Q&A] [answered]
### Question

I need to know whch host/IP is used to connect to server by the client.

This example works _almost_ fine, except that `await client.connected()` causes that the "page_layout" is not rendered (only stuff rendered by view.create_document_editor is visible), if I comment the await line, the page layout works.

```python
@ui.page('/cms/{app_id}/doc/')
@ui.page('/cms/{app_id}/doc/{document_id}')
async def cms_app_document_edit(client: Client, app_id, document_id=None):
    await client.connected()
    server_ip = client.environ['asgi.scope']['server'][0]
    with theme.page_layout('CMS', app.mqtt_client):
        view = cms_views.CmsEditor(app_id)
        view.create_document_editor(document_id)
```

**Comment by rodja** [comment]
Does a simple label show up?

```py
@ui.page('/cms/{app_id}/doc/')
@ui.page('/cms/{app_id}/doc/{document_id}')
async def cms_app_document_edit(client: Client, app_id, document_id=None):
    await client.connected()
    ui.label('cms content')
```

**Reply by marcin-koziol** [reply]
Yes.
Also, when I replace my other stuff inside the context block with simple label it still doesn't render the outer layout.

**Reply by rodja** [reply]
Wow. Then let's boil it down to the minimum!

```py
@ui.page('/')
async def index(client: Client):
    await client.connected()
    ui.label('cms content')
```

This MUST work. Otherwise a lot of the NiceGUI features would be broken...

**Reply by marcin-koziol** [reply]
Yes, this is OK, but why it pops out the page_layout context when using client.connected() ?

**Reply by rodja** [reply]
> Yes, this is OK

Thats a relief :-)

> but why it pops out the page_layout context when using client.connected() ?

I do not understand. What do you mean with "pops out"?

I re-read the conversation up to this point and now understand that this code also shows the label:

```py
@ui.page('/cms/{app_id}/doc/')
@ui.page('/cms/{app_id}/doc/{document_id}')
async def cms_app_document_edit(client: Client, app_id, document_id=None):
    await client.connected()
    ui.label('cms content')
```

That means something must be wrong with the context manager of `theme.page_layout('CMS', app.mqtt_client):`. Please try to remove the `app.mqtt_client` dependency. And then further reduce `theme.page_layout` to the bare minimum which reproduces the issue. This stripped-down version would be very interesting to see. Please share!

**Reply by marcin-koziol** [reply]
ok,  a moron here, haven't noticed this beautiful error message:

```
Traceback (most recent call last):
  File "/home/avl/.pyenv/versions/avl-gui/lib/python3.11/site-packages/nicegui/background_tasks.py", line 52, in _handle_task_result
    task.result()
  File "/home/avl/.pyenv/versions/avl-gui/lib/python3.11/site-packages/nicegui/page.py", line 106, in wait_for_result
    return await result
           ^^^^^^^^^^^^
  File "/home/avl/system/avl-gui/main.py", line 142, in cms_app_document_edit
    with theme.page_layout('CMS', app.mqtt_client):
  File "/home/avl/.pyenv/versions/3.11.6/lib/python3.11/contextlib.py", line 144, in __exit__
    next(self.gen)
  File "/home/avl/system/avl-gui/theme.py", line 57, in page_layout
    with ui.header(elevated=False).style('align-items: center').classes('dark:bg-gray-900'):
         ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/avl/.pyenv/versions/avl-gui/lib/python3.11/site-packages/nicegui/page_layout.py", line 62, in __init__
    add_body_html(f'''
  File "/home/avl/.pyenv/versions/avl-gui/lib/python3.11/site-packages/nicegui/functions/html.py", line 35, in add_body_html
    raise RuntimeError('Cannot add body HTML after the page has been sent to the client.')
RuntimeError: Cannot add body HTML after the page has been sent to the client.
Exception in callback <function page.__call__.<locals>.check_for_late_return_value at 0x7ff1948cb740>
handle: <Handle page.__call__.<locals>.check_for_late_return_value>
Traceback (most recent call last):
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/home/avl/.pyenv/versions/avl-gui/lib/python3.11/site-packages/nicegui/page.py", line 92, in check_for_late_return_value
    if task.result() is not None:
       ^^^^^^^^^^^^^
RuntimeError: Cannot add body HTML after the page has been sent to the client.
```

soo... here's my original question: how to get hostname of the server properly and not breaking other stuff?

**Reply by retsyo** [reply]
please read https://github.com/zauberzeug/nicegui/discussions/2140

but sometimes it may be fixed by moving all `add_body_html` before you create any element, sometimes you have to add `add_body_html` to the up, or upper function. English is not my native language, I hope I have expressed the meaning correctly.

**Comment by falkoschindler** [comment, marked as answer]
Is this what you're looking for?

```py
from fastapi import Request

@ui.page('/')
def page(request: Request):
    print(request.url)
```
